theano/tensor/nnet/nnet.py

```diff
@@ -465,119 +462,114 @@ def c_code_template(dtype):
         # TODO: put this into a templated function, in the support code
         # TODO: declare the max of each row as an Op output
 
-        # TODO: set error messages for failures in this code
-
         # TODO: use this to accept float32 and int32: node.inputs[0].type.dtype_specs()[1]
         init_decl = """
-        npy_intp* Nx = PyArray_DIMS(%(x)s);
-        npy_intp Sx1 = 0;
-        npy_intp Ssm1 = 0;
+            npy_intp* input_shape = PyArray_DIMS(%(x)s);
+            npy_int ndim = PyArray_NDIM(%(x)s);
 
-        if (PyArray_NDIM(%(x)s) != 2)
-        {
-            PyErr_SetString(PyExc_ValueError, "not a 2d tensor");
-            %(fail)s;
-        }
-        if ((PyArray_TYPE(%(x)s) != NPY_DOUBLE) &&
-            (PyArray_TYPE(%(x)s) != NPY_FLOAT))
-        {
-            PyErr_SetString(PyExc_TypeError, "not a float");
-            %(fail)s;
-        }
+            // Check input types
+            if ((PyArray_TYPE(%(x)s) != NPY_DOUBLE) &&
+                (PyArray_TYPE(%(x)s) != NPY_FLOAT))
+            {
+                PyErr_SetString(PyExc_TypeError, "Not a float");
+                %(fail)s;
+            }
 
-        if ((NULL == %(sm)s)
-            || (PyArray_DIMS(%(sm)s)[0] != PyArray_DIMS(%(x)s)[0])
-            || (PyArray_DIMS(%(sm)s)[1] != PyArray_DIMS(%(x)s)[1]))
-        {
-            Py_XDECREF(%(sm)s);
-            %(sm)s = (PyArrayObject*)PyArray_SimpleNew(2, PyArray_DIMS(%(x)s),
-                                                       PyArray_TYPE(%(x)s));
-            if(!%(sm)s) {
-                PyErr_SetString(PyExc_MemoryError,
-                     "failed to alloc sm output");
-                %(fail)s
+            // Allocate memory for the softmax output
+            if (NULL == %(sm)s || (PyArray_DIMS(%(x)s) != PyArray_DIMS(%(sm)s)))
```

Comment:
@lamblin. Do we need to check each dimension separately or we can check it like that. I can't think of a case where ```x.shape != sm.shape```. Could you give an example? Then we can also add it to the tests.